File: examples/geometry_viewer.cpp
// File: geometry_viewer.cpp
// Project: examples
// Created Date: 28/09/2022
// Author: Shun Suzuki
// -----
// Last Modified: 17/10/2022
// Modified By: Shun Suzuki (suzuki@hapis.k.u-tokyo.ac.jp)
// -----
// Copyright (c) 2022 Shun Suzuki. All rights reserved.
//

#include "autd3/extra/geometry_viewer.hpp"

#include <iostream>

#include "autd3.hpp"

int main() try {
  autd3::Controller autd;

  autd.geometry().add_device(autd3::Vector3::Zero(), autd3::Vector3::Zero());                                                // bottom
  autd.geometry().add_device(autd3::Vector3(0, 0, autd3::DEVICE_WIDTH), autd3::Vector3(0, autd3::pi / 2.0, 0));              // left
  autd.geometry().add_device(autd3::Vector3(autd3::DEVICE_WIDTH, 0, autd3::DEVICE_WIDTH), autd3::Vector3(0, autd3::pi, 0));  // top
  autd.geometry().add_device(autd3::Vector3(autd3::DEVICE_WIDTH, 0, 0), autd3::Vector3(0, -autd3::pi / 2.0, 0));             // right

  autd3::extra::GeometryViewer().window_size(800, 600).vsync(true).model(AUTD3_GEOMETRY_VIEWER_MODEL_PATH).view(autd.geometry());

  return 0;
} catch (std::exception& e) {
  std::cerr << e.what() << std::endl;
  return -1;
}

File: examples/src/geometry_viewer.rs
/*
 * File: simulator_server copy.rs
 * Project: src
 * Created Date: 24/05/2023
 * Author: Shun Suzuki
 * -----
 * Last Modified: 27/05/2023
 * Modified By: Shun Suzuki (suzuki@hapis.k.u-tokyo.ac.jp)
 * -----
 * Copyright (c) 2023 Shun Suzuki. All rights reserved.
 *
 */

use autd3::prelude::*;
use autd3_geometry_viewer::GeometryViewer;

fn main() {
    let geometry = Geometry::builder()
        .add_device(AUTD3::new(Vector3::zeros(), Vector3::zeros()))
        .add_device(AUTD3::new(
            Vector3::new(0., 0., DEVICE_WIDTH),
            Vector3::new(0., PI / 2., 0.),
        ))
        .add_device(AUTD3::new(
            Vector3::new(DEVICE_WIDTH, 0., DEVICE_WIDTH),
            Vector3::new(0., PI, 0.),
        ))
        .add_device(AUTD3::new(
            Vector3::new(DEVICE_WIDTH, 0., 0.),
            Vector3::new(0., -PI / 2., 0.),
        ))
        .build()
        .unwrap();

    let res = GeometryViewer::new().run(&geometry);
    std::process::exit(res);
}
